### ISSUE TITLE
How do I use this package to pass analytical jvp?

### ISSUE THREAD

**Issue by rveltz** [SciML/SciMLOperators.jl#223]
Hi,

Say I have the following vector field:

```julia
f(du,u,p,t) = @. du = -u^2
```

I know that the jacobian is given by

```
myjac(out,in,u,p,t) = @. out = -2 * u * in
```

How do I pass this information to `ODEFunction` or `ODEProblem`?

`SciMLOperators.FunctionOperator` is just asking for the linear op in the form `op(out, in, p, t)`, but where is the information about  `u`?

**Comment by ChrisRackauckas** [comment]
@avik-pal you just took a look through this could you comment?

**Comment by avik-pal** [comment]
`myjac(out,in,u,p,t) = @. out = -2 * u * in`

What is `in`? This looks like the JVP / VJP and not the Jacobian itself.

**Comment by rveltz** [comment]
yes that the jvp

**Comment by avik-pal** [comment]
Right, see how the `JacVec` operator is defined. Instead of using ad for the internal `op`, you will have to specify it analytically for every `u`. The other option is the use the `kwargs` interface for FunctionOperator and pass in `in` as a kwarg, for this see the `VecJac` implementation. Both are in https://github.com/JuliaDiff/SparseDiffTools.jl/blob/master/src/differentiation/jaches_products.jl & https://github.com/JuliaDiff/SparseDiffTools.jl/blob/master/src/differentiation/vecjac_products.jl

**Comment by rveltz** [comment]
I thought DE would provide a built in solution like DiffEqOperator used to.

**Comment by ChrisRackauckas** [comment]
This seems like a mistake. You could use parameters here, but it feels off. The function operator is `L(u,p,t)` so that it defines `mul!(y,L,x)` to be `y = L(u,p,t)*x`. The `u` seems to just be lopped off here. You can do `L((u,p),t)*x` but that seems like a mistake. `input` isn't `u`: you're not necessarily always doing `L(u,p,t)*u`, so the state of the operator needs to update independent of the vector that is being multiplied.

Wouldn't this make JacVec easier to implement too?

**Comment by ChrisRackauckas** [comment]
We talked, FunctionOperator is just wrong. We need to fix that.

**Comment by rveltz** [comment]
👍

**Comment by vpuri3** [comment]
We have come across this issue before when trying to form nonlinear operators eg, `f(u) * u, df(u)/du * u`, etc with ComposedOperator (https://github.com/SciML/SciMLOperators.jl/issues/159). This seemed like a design constraint for `SciMLOperators`, and so when writing `SparseDiffTools.JacVec/VecJac`, we created the `kwarg` interface for `FunctionOperator` (https://github.com/SciML/SciMLOperators.jl/issues/125), That would allow calls to `VecJac` like `L(v, u, p, t; VJP_input = w)` which is equivalent to `v .= df(w)/dw * u` (https://github.com/JuliaDiff/SparseDiffTools.jl/pull/245). I don't know the current problem you're facing, but maybe a similar solution for `JacVec` could help alleviate the problem?

**Comment by ChrisRackauckas** [comment]
> That would allow calls to VecJac like L(v, u, p, t; VJP_input = w) which is equivalent to v .= df(w)/dw * u (https://github.com/JuliaDiff/SparseDiffTools.jl/pull/245).

Let's standardize our notation.

```
w = L(u,p,t)*v
```

**Comment by ChrisRackauckas** [comment]
The `L` someone would define here is just `L = FunctionOperator((w,v,u,p,t)-> ..., u, p, t)`, where `update_coefficients!(L,u,p,t)` updates the internal states.

**Comment by ChrisRackauckas** [comment]
L should have two calls:

- `L(v,u,p,t)` out of place
- `L(w,v,u,p,t)` in place

**Comment by rveltz** [comment]
> Let's standardize our notation.
 
 If I understand you well, `L(u,p,t)` is a linear operator representing the  differential of say `F(u,p,t)` wrt to the first variable at position `u`.

**Comment by vpuri3** [comment]
I'm tied down with school work at the moment. I have a long flight this weekend when I'm gonna start working on this.
although it's a small change it's gonna require releasing v0.4 and subsequent modifications in downstream packages.

**Comment by rveltz** [comment]
Sorry to be pushy... did  you make any progress?

**Comment by ChrisRackauckas** [comment]
No. But, CI in a few places has been blocked by https://github.com/bifurcationkit/BifurcationKit.jl/pull/127 so while I have your attention if that can get fixed that would help a lot.

**Comment by rveltz** [comment]
bump ;)

**Comment by avik-pal** [comment]
The next release of NonlinearSolve will have a `JacobianOperator` which could be a reference for the new FunctionOperator re-design https://github.com/SciML/NonlinearSolve.jl/blob/ap/rework_internals/src/internal/operators.jl

**Comment by rveltz** [comment]
Have we made progress on this side? it is still unusable for me.

**Comment by rveltz** [comment]
I see that you added the operators. Can you show me an example how to solve the current issue please?

**Comment by ChrisRackauckas** [comment]
No progress has been made. This is probably the package in the worst state right now. I need to find time for it.

**Comment by HumHongeKamyaab** [comment]
>  I want to directly specify user-defined Jacobian vector product (JVP) for my system `jvp(Jv,v,u,p,t)`

Hi @ChrisRackauckas, Is there any update on this issue, or Is there any alternative way without using FunctionalOperator to specify JVP of the system so that Krylov.jl can use it in `mul!` via DifferencialEquation.jl.

**Comment by avik-pal** [comment]
We now have https://docs.sciml.ai/NonlinearSolve/dev/api/SciMLJacobianOperators/ but currently only has bindings if you specify it is a nonlinearproblem

**Comment by HumHongeKamyaab** [comment]
@avik-pal  Thanks. 
I am not sure if can solve for problem defined using NonlinearSolve during an impilcit ODE step of DifferencialEquation.jl
 I will check for test examples if it is possible.

**Comment by ChrisRackauckas** [comment]
This is fixed in https://github.com/SciML/SciMLOperators.jl/pull/269 and thus SciMLOperators v1 has the fix.

We will update SciMLJacobianOperators.jl to be a sublib here that defines jvp/vjp operators on v1.

**Comment by rveltz** [comment]
Yes!
